src/ui/win/win_ui.cpp

```diff
@@ -0,0 +1,257 @@
+    {
+        switch ( msg ) {
+            case WM_CREATE:
+                // register owner class object pointer
+                SetWindowLongPtr( window, GWLP_USERDATA, reinterpret_cast<LONG_PTR>(reinterpret_cast<CREATESTRUCT*>(lp)->lpCreateParams) );
+                break;
+            case WM_PAINT:
+            {
+                PAINTSTRUCT paintStructure;
+                HDC handleToDisplayDevice = BeginPaint( window, &paintStructure );
+
+                const WindowsUi::UiWindowWinInfo info( reinterpret_cast<UiWindowWin*>(GetWindowLongPtr( window, GWLP_USERDATA )) );
+                if ( info.valid() ) { // make sure that we draw a valid image
+                    RECT clientRoi;
+                    GetClientRect( window, &clientRoi );
+                    const DWORD result = StretchDIBits( handleToDisplayDevice, 0, 0, clientRoi.right, clientRoi.bottom,
+                                                        0, 0, info.image().width(), info.image().height(),
+                                                        info.image().data(), info.bitmapInfo(), DIB_RGB_COLORS, SRCCOPY );
+                    if ( result != info.image().height() )
+                        DebugBreak(); // drawing failed
+
+                    info.paint( handleToDisplayDevice, clientRoi );
+                }
+                EndPaint( window, &paintStructure );
+            }
+            break;
+            case WM_DESTROY:
+                PostQuitMessage( 0 ) ;
+            default:
+                return DefWindowProc( window, msg, wp, lp );
+        }
+
+        return 0;
+    }
+
+    struct UiWindowWinRegistrator
+    {
+        UiWindowWinRegistrator()
+            : registered( 0 )
+            , className( L"UiWindowWin" )
+        {
+            WNDCLASSEX wcex;
+            wcex.cbSize        = sizeof( WNDCLASSEX );
+            wcex.style         = CS_HREDRAW | CS_VREDRAW;
+            wcex.lpfnWndProc   = WindowProcedure;
+            wcex.cbClsExtra    = 0;
+            wcex.cbWndExtra    = 0;
+            wcex.hInstance     = 0;
+            wcex.hIcon         = 0;
+            wcex.hCursor       = LoadCursor( NULL, IDC_CROSS );
+            wcex.hbrBackground = reinterpret_cast<HBRUSH>(COLOR_WINDOW);
+            wcex.lpszMenuName  = NULL;
+            wcex.lpszClassName = className.data();
+            wcex.hIconSm       = 0;
+
+            registered = RegisterClassEx( &wcex );
+        }
+        WORD registered;
+        const std::wstring className;
+    };
+}
+
+UiWindowWin::UiWindowWin( const PenguinV_Image::Image & image, const std::string & title )
+    : UiWindow( PenguinV_Image::Image(), title ) // we pass an empty image into base class
+    , _bmpInfo( nullptr )
+{
+    static const UiWindowWinRegistrator registrator; // we need to register only once hence static variable
+    if ( !registrator.registered )
+        throw imageException( "Unable to create Windows API class" );
+
+    setImage( image );
```

Comment:
Calling virtual function in ctor is dangerous. https://isocpp.org/wiki/faq/strange-inheritance#calling-virtuals-from-ctors

Reply:
Okay, we can do like this:
```cpp
UiWindowWin::setImage( image )
```